File: android/app/src/main/java/com/demo/videopick/viewmodel/DownloadViewModel.kt
package com.demo.videopick.viewmodel

import android.app.Application
import android.content.ClipboardManager
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.demo.videopick.data.model.MediaType
import com.demo.videopick.data.model.VideoInfo
import com.demo.videopick.data.repository.DownloadRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private const val CLIPBOARD_HINT_TEXT = "已识别到剪贴板内容，已自动填入"

data class UiState(
    val inputText: String = "",
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val successMessage: String? = null,
    val videoInfo: VideoInfo? = null,
    val showPreview: Boolean = false,
    val downloadProgress: Float? = null,
    val clipboardHint: String? = null,
)

class DownloadViewModel(application: Application) : AndroidViewModel(application) {

    private val repository = DownloadRepository(application)
    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state.asStateFlow()

    private var downloadJob: Job? = null
    private var lastClipboardContent: String? = null
    private var clipboardHintDismissJob: Job? = null

    fun updateInput(text: String) {
        _state.value = _state.value.copy(inputText = text)
    }

    fun clearInput() {
        _state.value = UiState()
    }

    fun processInput() {
        val text = _state.value.inputText.trim()
        if (text.isEmpty()) {
            _state.value = _state.value.copy(errorMessage = "请输入分享链接")
            return
        }
        downloadJob = viewModelScope.launch {
            download(text)
        }
    }

    fun cancelDownload() {
        downloadJob?.cancel()
        downloadJob = null
        _state.value = _state.value.copy(
            isLoading = false,
            downloadProgress = null,
        )
    }

    private suspend fun download(text: String) {
        _state.value = _state.value.copy(
            isLoading = true,
            errorMessage = null,
            successMessage = null,
            videoInfo = null,
            showPreview = false,
            downloadProgress = null,
        )

        try {
            val info = withContext(Dispatchers.IO) {
                repository.parseAndDownload(text) { progress ->
                    _state.value = _state.value.copy(downloadProgress = progress)
                }
            }
            _state.value = _state.value.copy(
                videoInfo = info,
                showPreview = true,
            )
        } catch (e: Exception) {
            _state.value = _state.value.copy(
                errorMessage = e.message ?: "未知错误",
            )
        } finally {
            _state.value = _state.value.copy(
                isLoading = false,
                downloadProgress = null,
            )
        }
    }

    /** Story 8.1.1: 前台切换时自动读取剪贴板，若内容变化则填入并提示。 */
    fun checkClipboardOnForeground() {
        // AC13: 首次（cache 为 null）时把已有 inputText 同步进 cache，避免 share intent 等预填导致重复触发
        if (lastClipboardContent == null && _state.value.inputText.isNotEmpty()) {
            lastClipboardContent = _state.value.inputText
        }

        val context = getApplication<Application>().applicationContext
        val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val primaryClip = clipboard.primaryClip ?: return
        if (primaryClip.itemCount == 0) return
        val clipboardString = primaryClip.getItemAt(0).coerceToText(context).toString()

        // AC3: 三条件 AND（非空 + 与缓存不同）
        if (clipboardString.isEmpty()) return
        if (clipboardString == lastClipboardContent) return

        // AC6: 缓存先更新；AC7: 替换 inputText
        lastClipboardContent = clipboardString
        _state.value = _state.value.copy(inputText = clipboardString)
        showClipboardHint(CLIPBOARD_HINT_TEXT)
    }

    private fun showClipboardHint(message: String) {
        // AC9: cancel-and-restart，不堆叠
        clipboardHintDismissJob?.cancel()
        _state.value = _state.value.copy(clipboardHint = message)
        clipboardHintDismissJob = viewModelScope.launch {
            delay(3000)
            _state.value = _state.value.copy(clipboardHint = null)
        }
    }

    fun saveMedia() {
        val info = _state.value.videoInfo ?: return
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, errorMessage = null, successMessage = null)
            try {
                val msg = withContext(Dispatchers.IO) {
                    when (info.mediaType) {
                        MediaType.VIDEO -> {
                            val path = info.localPath ?: throw Exception("没有可保存的视频")
                            repository.saveVideoToGallery(path)
                        }
                        MediaType.IMAGES -> {
                            if (info.localImagePaths.isEmpty()) throw Exception("没有可保存的图片")
                            repository.saveImagesToGallery(info.localImagePaths)
                        }
                    }
                }
                _state.value = _state.value.copy(successMessage = msg)
            } catch (e: Exception) {
                _state.value = _state.value.copy(errorMessage = e.message ?: "保存失败")
            } finally {
                _state.value = _state.value.copy(isLoading = false)
            }
        }
    }
}

File: flutter/lib/viewmodels/download_viewmodel.dart
import 'dart:async';
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:photo_manager/photo_manager.dart';
import '../models/video_info.dart';
import '../services/download_service.dart';

class DownloadViewModel extends ChangeNotifier with WidgetsBindingObserver {
  static const String kClipboardHintText = '已识别到剪贴板内容，已自动填入';

  final _service = DownloadService();

  String inputText = '';
  bool isLoading = false;
  String? errorMessage;
  String? successMessage;
  VideoInfo? videoInfo;
  bool showPreview = false;
  double? downloadProgress;
  String? clipboardHint;

  Completer<void>? _cancelCompleter;
  bool _isCancelled = false;
  String? _lastClipboardContent;
  Timer? _clipboardHintDismissTimer;

  DownloadViewModel() {
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    _clipboardHintDismissTimer?.cancel();
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    if (state == AppLifecycleState.resumed) {
      checkClipboardOnForeground();
    }
  }

  /// Story 8.1.2: 前台切换时自动读取剪贴板，若内容变化则填入并提示。
  Future<void> checkClipboardOnForeground() async {
    // AC13: 首次（cache 为 null）时把已有 inputText 同步进 cache
    if (_lastClipboardContent == null && inputText.isNotEmpty) {
      _lastClipboardContent = inputText;
    }

    final data = await Clipboard.getData(Clipboard.kTextPlain);
    final text = data?.text;
    if (text == null || text.isEmpty) return;
    if (text == _lastClipboardContent) return;

    // AC6: 缓存先更新；AC7: 替换 inputText 并 notify
    _lastClipboardContent = text;
    inputText = text;
    notifyListeners();
    _showClipboardHint(kClipboardHintText);
  }

  void _showClipboardHint(String message) {
    // AC9: cancel-and-restart，不堆叠
    _clipboardHintDismissTimer?.cancel();
    clipboardHint = message;
    notifyListeners();
    _clipboardHintDismissTimer = Timer(const Duration(seconds: 3), () {
      clipboardHint = null;
      notifyListeners();
    });
  }

  void updateInput(String text) {
    inputText = text;
    notifyListeners();
  }

  void clearInput() {
    inputText = '';
    errorMessage = null;
    successMessage = null;
    videoInfo = null;
    showPreview = false;
    downloadProgress = null;
    notifyListeners();
  }

  Future<void> pasteFromClipboard() async {
    final data = await Clipboard.getData(Clipboard.kTextPlain);
    if (data?.text != null && data!.text!.isNotEmpty) {
      inputText = data.text!;
      notifyListeners();
    }
  }

  Future<void> processInput() async {
    if (inputText.trim().isEmpty) return;

    isLoading = true;
    errorMessage = null;
    successMessage = null;
    videoInfo = null;
    showPreview = false;
    downloadProgress = null;
    _isCancelled = false;
    _cancelCompleter = Completer<void>();
    notifyListeners();

    try {
      final info = await _service.parseAndDownload(
        inputText,
        onProgress: (progress) {
          if (_isCancelled) return;
          downloadProgress = progress;
          notifyListeners();
        },
      );

      if (_isCancelled) return;

      videoInfo = info;
      showPreview = true;
      downloadProgress = null;
    } catch (e) {
      if (_isCancelled) return;
      errorMessage = e.toString();
    } finally {
      isLoading = false;
      notifyListeners();
    }
  }

  void cancelDownload() {
    _isCancelled = true;
    _cancelCompleter?.complete();
    isLoading = false;
    downloadProgress = null;
    notifyListeners();
  }

  Future<void> saveMedia() async {
    if (videoInfo == null) return;

    final permission = await PhotoManager.requestPermissionExtend();
    if (!permission.isAuth) {
      errorMessage = '没有相册访问权限，请在设置中开启';
      notifyListeners();
      return;
    }

    try {
      if (videoInfo!.mediaType == MediaType.images) {
        for (final path in videoInfo!.localImagePaths) {
          final file = File(path);
          if (await file.exists()) {
            await PhotoManager.editor.saveImageWithPath(
              path,
              title: 'VideoPick_${DateTime.now().millisecondsSinceEpoch}',
            );
          }
        }
        successMessage = '已保存 ${videoInfo!.localImagePaths.length} 张图片到相册';
      } else {
        final localPath = videoInfo!.localPath;
        if (localPath != null) {
          final file = File(localPath);
          if (await file.exists()) {
            await PhotoManager.editor.saveVideo(
              file,
              title: 'VideoPick_${DateTime.now().millisecondsSinceEpoch}',
            );
            successMessage = '视频已保存到相册';
          }
        }
      }
    } catch (e) {
      errorMessage = '保存失败: $e';
    }

    notifyListeners();
  }

  void dismissPreview() {
    showPreview = false;
    notifyListeners();
  }
}
